src/responder/kcm/kcmsrv_op_queue.c

```diff
@@ -0,0 +1,230 @@
+/*
+   SSSD
+
+   KCM Server - the KCM operations wait queue
+
+   Copyright (C) Red Hat, 2017
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "util/util.h"
+#include "util/util_creds.h"
+#include "responder/kcm/kcmsrv_pvt.h"
+
+#define QUEUE_HASH_SIZE      32
+
+struct kcm_ops_queue_entry {
+    struct tevent_req *req;
+    uid_t uid;
+
+    hash_table_t *wait_queue_hash;
+
+    struct kcm_ops_queue_entry **head;
+    struct kcm_ops_queue_entry *next;
+    struct kcm_ops_queue_entry *prev;
+};
+
+struct kcm_ops_queue_ctx {
+    /* UID: dlist of kcm_ops_queue_entry */
+    hash_table_t *wait_queue_hash;
+};
+
+/*
+ * Per-UID wait queue
+ */
+struct kcm_ops_queue_ctx *kcm_ops_queue_create(TALLOC_CTX *mem_ctx)
+{
+    errno_t ret;
+    struct kcm_ops_queue_ctx *queue_ctx;
+
+    queue_ctx = talloc_zero(mem_ctx, struct kcm_ops_queue_ctx);
+    if (queue_ctx == NULL) {
+        return NULL;
+    }
+
+    ret = sss_hash_create_ex(mem_ctx, QUEUE_HASH_SIZE,
+                             &queue_ctx->wait_queue_hash, 0, 0, 0, 0,
+                             NULL, NULL);
+    if (ret != EOK) {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "sss_hash_create failed [%d]: %s\n", ret, sss_strerror(ret));
+        talloc_free(queue_ctx);
+        return NULL;
+    }
+
+    return queue_ctx;
+}
+
+static int kcm_op_queue_entry_destructor(struct kcm_ops_queue_entry *entry)
+{
+    int ret;
+    struct kcm_ops_queue_entry *next_entry;
+    hash_key_t key;
+
+    if (entry == NULL) {
+        return 1;
+    }
+
+    /* Take the next entry from the queue */
+    next_entry = entry->next;
+
+    /* Remove the current entry from the queue */
+    DLIST_REMOVE(*entry->head, entry);
+
+    if (next_entry == NULL) {
```

Comment:
The hash's key is the UID of the client, the value is the queue. If there is nothing in the queue at the time the request arrives (`hash[uid] == none`) then we create a request and add it to the hash table (`hash[uid] = new_request()`) but continue immediatelly. If there is already some request (`if uid in hash`) then we append the request (`hash[uid].append(new_request())`) and wait for the previous requests to finish.

I commented the code more and also added a unit test for the module, hopefully it's clearer now.